### PR TITLE
Madina_Dzhetegenova_Ada_C22_Phoenix

### PR DESCRIPTION
**Pull request by Madina-j**

### CODE REVIEW

File: app/routes/task_routes.py [hunk 1]
```diff
@@ -1 +1,177 @@
...
+        return {
+            "task": {
+                "id": task.id,
+                "title": task.title,
+                "description": task.description,
+                "is_complete": task.completed_at is not None
+            }
+        }
+
+def validate_task(task_id):
```
Comment (anselrognlie) [inline app/routes/task_routes.py:89]: 👀 Note that we could refactor this `validate_task` and the `validate_goal` functions to be able to handle both.

File: app/routes/goal_routes.py [hunk 2]
```diff
@@ -1 +1,145 @@
...
+def get_one_goal(goal_id):
+    goal = validate_goal(goal_id)
+
+    return { "goal" : {
+                "id": goal.id,
+                "title": goal.title
+            }
+    }
+
+def validate_goal(goal_id):
```
Comment (anselrognlie) [inline app/routes/goal_routes.py:57]: 👀 Note that we could refactor this `validate_goal` and the `validate_task` functions to be able to handle both.

File: tests/test_wave_01.py [hunk 3]
```diff
@@ -51,22 +51,22 @@ def test_get_task(client, one_task):
...
 def test_get_task_not_found(client):
     # Act
     response = client.get("/tasks/1")
     response_body = response.get_json()
 
     # Assert
     assert response.status_code == 404
-
-    raise Exception("Complete test with assertion about response body")
+    assert response_body == {"message": f"task 1 not found"}
```
Comment (anselrognlie) [inline tests/test_wave_01.py:62]: 👀 This message value doesn't need to be an f-string. There's no interpolation happening there.

File: tests/test_wave_01.py [hunk 4]
```diff
@@ -130,14 +130,14 @@ def test_update_task_not_found(client):
...
```
Comment (anselrognlie) [inline tests/test_wave_01.py:133]: Here again and throughout, the f-string is unnecessary.

File: tests/test_wave_05.py [hunk 5]
```diff
@@ -80,34 +82,48 @@ def test_create_goal(client):
...
     # ---- Complete Act Here ----
 
     # Assert
     # ---- Complete Assertions Here ----
-    # assertion 1 goes here
-    # assertion 2 goes here
-    # assertion 3 goes here
+    goal = Goal.query.get(1)
+    assert response.status_code == 200
+    assert goal.title == "Updated Title"
```
Comment (anselrognlie) [inline tests/test_wave_05.py:99]: Prefer to keep this line closer to line that retrieves the updated goal. It's a little confusing to have it mixed in with the response checks. But great idea to check that the attempted change is actually visible in the database (not just in the response).

File: tests/test_wave_05.py [hunk 6]
```diff
@@ -123,28 +139,31 @@ def test_delete_goal(client, one_goal):
     # Check that the goal was deleted
     response = client.get("/goals/1")
     assert response.status_code == 404
-
-    raise Exception("Complete test with assertion about response body")
+    assert response_body == {
+  "details": "Goal 1 \"Build a habit of going outside daily\" successfully deleted"
+}
```
Comment (anselrognlie) [inline tests/test_wave_05.py:144]: 👀 The `response_body` hasn't been updated to get the "missing" message from this second API call. That's why we're still seeing the first response. We would need to add something like
```py
    response_body = response.get_json()
```

We should make sure that what we're actually asserting makes sense for the case, and not just pass the test. Here, if we try to retrieve the goal we just deleted, we shouldn't be getting the delete success message again.

File: migrations/versions/251ac0f157f8_.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+"""empty message
```
Comment (anselrognlie) [inline migrations/versions/251ac0f157f8_.py:1]: Be sure to include a -m "message" option when migrating to add a note to the migration. This can make it easier to refer back to the migration later and determine what it's doing. Often, since the migrations are named first with a hash, they don't list in the order they were created, making it difficult to trace through a series of migrations.

Comment (anselrognlie) [reply migrations/versions/251ac0f157f8_.py:1]: It looks like you went through a couple of steps to create your migrations. This is completely fine, and reflects the progress of the work.

On the other hand, while I'm still in "development phase", I often like to have just one clean migration to setup the initial database. So another approach would be to periodically clear out the database and migrations, and make a fresh one once you've got the models configured as desired. This isn't a good idea once a project has been deployed publicly (since we would lose any exisiting data when we drop the database), but can make the migrations a little easier to understand for setting up a fresh deployment.

File: app/models/task.py [hunk 8]
```diff
@@ -1,5 +1,14 @@
...
+from sqlalchemy import ForeignKey
 class Task(db.Model):
     id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
+    title: Mapped[str]
+    description: Mapped[str]
+    completed_at: Mapped[Optional[datetime]]
+    goal_id: Mapped[Optional[int]] = mapped_column(ForeignKey("goal.id"))
+    goal: Mapped[Optional["Goal"]] = relationship(back_populates="tasks")
+
+
```
Comment (anselrognlie) [inline app/models/task.py:14]: Remember that we can add helper methods to do common actions like converting to dictionaries, or load from dictionaries.

File: app/models/goal.py
Comment (anselrognlie) [inline app/models/goal.py:1]: Like with the Task model, we can add helper methods to do common actions like converting to or loading from dictionaries.

File: app/__init__.py [hunk 9]
```diff
@@ -1,13 +1,16 @@
 from flask import Flask
 from .db import db, migrate
 from .models import task, goal
+from .routes.task_routes import tasks_bp
+from .routes.goal_routes import goals_bp
```
Comment (anselrognlie) [inline app/__init__.py:5]: We would prefer to name the `task_bp` and `goal_bp` just `bp` in their respective route files, and then either do an `import as` here to differentiate them, or import each entire route module and access its Blueprint using dor notation.

File: app/__init__.py [hunk 10]
```diff
@@ -1,13 +1,16 @@
 from flask import Flask
 from .db import db, migrate
 from .models import task, goal
```
Comment (anselrognlie) [inline app/__init__.py:3]: Once we have other import paths to our models (here, through the blueprints, which import the models), we technically don't need the imports here any more. It's fine to leave them for clarity (and a reminder that any other models we add would need to be included until other routes are setup), but if the VS Code warning is bothersome, feel free to remove these.

File: app/routes/__init__.py
Comment (anselrognlie) [inline app/routes/__init__.py:1]: Nice job remembering to create the `__init__.py`. Even if things appear to work without it, there are corner cases where it can become a problem, so we should always make it.

File: app/routes/task_routes.py [hunk 11]
```diff
@@ -1 +1,177 @@
...
+from flask import Blueprint, abort, make_response, request, Response
+from app.models.task import Task
+from app.models.goal import Goal
+from datetime import datetime
+from ..db import db
+import requests
+import os
+
+
+tasks_bp = Blueprint("tasks_bp", __name__, url_prefix="/tasks")
```
Comment (anselrognlie) [inline app/routes/task_routes.py:10]: Moving forward, prefer to name the Blueprint in a route file as just `bp` rather than having the variable include the model name. The string name would still be written as is for differentiation from other blueprints.

File: app/routes/task_routes.py [hunk 12]
```diff
@@ -1 +1,177 @@
...
+
+@tasks_bp.post("")
+def create_task():
+    request_body = request.get_json()
+    if "title" not in request_body:
+        return {"details": "Invalid data"}, 400
+    if "description" not in request_body:
+        return {"details": "Invalid data"}, 400
```
Comment (anselrognlie) [inline app/routes/task_routes.py:18]: Alternatively, we could watch for a `KeyError` when trying to retrieve these below. If we had a `from_dict` method, the key accesses could be within that method, and we have a `create_model` wrapper to watch for `KeyError`s and turn those into the appropriate error behavior.

File: app/routes/task_routes.py [hunk 13]
```diff
@@ -1 +1,177 @@
...
+
+    title = request_body["title"]
+    description = request_body["description"]
+    completed_at= request_body.get("completed_at")
```
Comment (anselrognlie) [inline app/routes/task_routes.py:22]: Nit: watch the spacing around assignment `=`

Comment (anselrognlie) [reply app/routes/task_routes.py:22]: Nice use of `get` to make the `"completed_at"` key be treated as optional.

File: app/routes/task_routes.py [hunk 14]
```diff
@@ -1 +1,177 @@
...
+
+    new_task = Task(title=title, description=description, completed_at= None if completed_at is None else datetime.datetime(completed_at)
```
Comment (anselrognlie) [inline app/routes/task_routes.py:24]: What kind of data is this expecting `completed_at` to have when it's non-`None`? The most common way to send datetime information through JSON is with a specially formatted string (there's no specific datetime type in JSON), but the python datetime [constructor](https://docs.python.org/3/library/datetime.html#datetime.datetime) doesn't accept strings. It expects the information to be passed in as several numbers. Instead, we might want to use the [datetime.fromisoformat()](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat) to read the data from a string.  Alternatively, depending on the string format, the database might be able to convert it from a string on its own.

There were no provided tests that passed in a `completed_at` value, so if you wanted to support being able to make a completed Task, be sure to write your own tests to check how the logic behaves.

File: app/routes/task_routes.py [hunk 15]
```diff
@@ -1 +1,177 @@
...
+    db.session.add(new_task)
+    db.session.commit()
+
+    
+    response = { "task" : {
+        "id": new_task.id,
+        "title": new_task.title,
+        "description": new_task.description,
+        "is_complete": new_task.completed_at is not None 
+    }
```
Comment (anselrognlie) [inline app/routes/task_routes.py:35]: Notice how the dictionary structure of a Task is needed in several of the endpoints (both here, and in the goals routes). Including a `to_dict` method in Task would allow this repetitive logic to be kept in a single place, with the guarantee that they are all generating the same output format.

File: app/routes/task_routes.py [hunk 16]
```diff
@@ -1 +1,177 @@
...
+@tasks_bp.get("")
+def get_all_tasks():
+    sort_order = request.args.get("sort")
+
+    if sort_order == "asc":
+        query = db.select(Task).order_by(Task.title.asc())
+    elif sort_order == "desc":
+        query = db.select(Task).order_by(Task.title.desc())
+    else:
+        query = db.select(Task).order_by(Task.id)
```
Comment (anselrognlie) [inline app/routes/task_routes.py:48]: Nice fallback to order by id when no sort is requested.

Comment (anselrognlie) [inline app/routes/task_routes.py:48]: Note that the `db.select` is repeated in each of these. Consider starting to build a query with the shared portion, then update it as needed according to any passed options.
```py
    query = db.select(Task)

    if sort_order == "asc":
        query = query.order_by(Task.title.asc())
    elif sort_order == "desc":
        query = query.order_by(Task.title.desc())
    else:
        query = query.order_by(Task.id)
```

Even though it doesn't shorten things much here, it's a good practice to get into.

File: app/routes/task_routes.py [hunk 17]
```diff
@@ -1 +1,177 @@
...
+        }
+    else:
+        return {
+            "task": {
+                "id": task.id,
+                "title": task.title,
+                "description": task.description,
+                "is_complete": task.completed_at is not None
+            }
+        }
```
Comment (anselrognlie) [inline app/routes/task_routes.py:87]: Notice that most of the keys are returned in either case. Only the goal id is variable. We could generate a dictionary holding the common values first, then add in the goal id if needed. Something like this.
```py
    result = {
            "id": task.id,
            "title": task.title,
            "description": task.description,
            "is_complete": task.completed_at is not None
        }

    if task.goal_id is not None:
        result["goal_id"] = task.goal_id

    return {"task": result}
```

File: app/routes/task_routes.py [hunk 18]
```diff
@@ -1 +1,177 @@
...
+        abort(make_response(response, 404))
+    return task
+
+@tasks_bp.put("/<task_id>")
+def update_task(task_id):
+    task = validate_task(task_id)
+    request_body = request.get_json()
+
+    task.title = request_body["title"]
+    task.description = request_body["description"]
```
Comment (anselrognlie) [inline app/routes/task_routes.py:110]: The same kind of key errors that were possible in the POST route could also happen here in the PUT, so this route could also benefit from similar error handling.

File: app/routes/task_routes.py [hunk 19]
```diff
@@ -1 +1,177 @@
...
+
+@tasks_bp.delete("/<task_id>")
+def delete_task(task_id):
+    task = validate_task(task_id)
+    
+    db.session.delete(task)
+    db.session.commit()
+
+    response = {
+        "details": f"Task {task.id} \"{task.title}\" successfully deleted"
```
Comment (anselrognlie) [inline app/routes/task_routes.py:133]: Since this string needs to store double quotes, we could use a single quote f-string to avoid having to escpae the quotes.

File: app/routes/task_routes.py [hunk 20]
```diff
@@ -1 +1,177 @@
...
+    }
+    
+    return response, 200
+
+@tasks_bp.patch("/<task_id>/mark_complete")
+def update_task_to_complete(task_id):
+    task = validate_task(task_id)
+    task.completed_at = datetime.now()
```
Comment (anselrognlie) [inline app/routes/task_routes.py:141]: `datetime` values are surpisingly difficult to work with correctly. This is why we only ask for a vlaue to be stored to mark completion, with the lack of a value indiciation incomplete. Especially for apps that might be storing information from remote users from around the world, a common approach for storing datetimes is to store timezone-aware data in the UTC timezone, then leave it to the local client to transform the value to the local timezone for display. This requires both using timezone-aware storage in the database, as well as timezone-aware Python calls. It's definitely worth looking into for future work!

File: app/routes/task_routes.py [hunk 21]
```diff
@@ -1 +1,177 @@
...
+        "Authorization": f"Bearer {token}" 
+    }
+    url = "https://slack.com/api/chat.postMessage"
+
+    message = {"channel": "D07V10LPBM4",
+            "text": f"Someone just completed the task {task.title}"}
+
+    response = requests.post(url, headers=headers, data=message)
+    
+    print(response.text)
```
Comment (anselrognlie) [inline app/routes/task_routes.py:154]: Consider moving this to a helper function so that the completion route can stay focused on marking the task complete, with the logic related to posting a message handled elsewhere.

File: app/routes/goal_routes.py
Comment (anselrognlie) [inline app/routes/goal_routes.py:1]: Check the equivalent routes in the task_routes file for points to watch out for here.

File: app/routes/goal_routes.py [hunk 22]
```diff
@@ -1 +1,145 @@
-from flask import Blueprint
\ No newline at end of file
+from flask import Blueprint, abort, make_response, request, Response
+from app.models.goal import Goal
+from app.models.task import Task
+from datetime import datetime
+from ..db import db
+from .task_routes import validate_task
+
+goals_bp = Blueprint("goals_bp", __name__, url_prefix="/goals")
```
Comment (anselrognlie) [inline app/routes/goal_routes.py:8]: Like for tasks, we would prefer to call this variable `bp` (no model in the name).

File: app/routes/goal_routes.py [hunk 23]
```diff
@@ -1 +1,145 @@
...
+    new_goal = Goal(title=title)
+    db.session.add(new_goal)
+    db.session.commit()
+
+    
+    response = { "goal" : {
+        "id": new_goal.id,
+        "title": new_goal.title
+    }
+    }
```
Comment (anselrognlie) [inline app/routes/goal_routes.py:29]: Like for tasks, a `to_dict` helper would reduce repetition in all the places where we need to generate a goal dictionary response.

File: app/routes/goal_routes.py [hunk 24]
```diff
@@ -1 +1,145 @@
...
+    goals = db.session.scalars(query)
+
+    goals_response = []
+    for goal in goals:
+        goals_response.append(
+            {
+                "id": goal.id,
+                "title": goal.title
+            }
+        )
```
Comment (anselrognlie) [inline app/routes/goal_routes.py:44]: With a `to_dict` method, a list comprehension would become a readable alternative to writing out the loop here longhand.
```py
    goals_response = [goal.to_dict() for goal in goals]
```

File: app/routes/goal_routes.py [hunk 25]
```diff
@@ -1 +1,145 @@
...
+        abort(make_response(response, 404))
+    return goal
+
+
+@goals_bp.put("/<goal_id>")
+def update_goal(goal_id):
+    goal = validate_goal(goal_id)
+    request_body = request.get_json()
+
+    goal.title = request_body["title"]
```
Comment (anselrognlie) [inline app/routes/goal_routes.py:78]: Similar to the Task PUT route, this route could benefit from some error handling for any missing keys.

File: app/routes/goal_routes.py [hunk 26]
```diff
@@ -1 +1,145 @@
...
+@goals_bp.post("/<goal_id>/tasks")
+def create_task_with_goal(goal_id):
+    goal = validate_goal(goal_id)
+
+    request_data = request.get_json()
+    task_ids = request_data.get("task_ids")
+
+    if not task_ids:
+        response = {"message": "No task_ids provided"}
+        abort(make_response(response, 400))
```
Comment (anselrognlie) [inline app/routes/goal_routes.py:114]: We don't include a test for passing an empty list here, so treating it as an error is fine. Consider adding additional test cases if you add additional logic not covered by existing tests so that you can try this out.

An alternative way to handle this might be to interpret sending an empty list as meaning to disassociate the goal from any current tasks, though this might work better if we make a different behavioral choice below.

File: app/routes/goal_routes.py [hunk 27]
```diff
@@ -1 +1,145 @@
...
+
+    for task_id in task_ids:
+        task = validate_task(task_id)
+        task.goal = goal
```
Comment (anselrognlie) [inline app/routes/goal_routes.py:118]: By setting the goal for each task to be the current goal, this route is effectively appending the supplied tasks to the goal. There might have been other tasks already associated with the goal, though such a scenario doesn't appear in any of the tests we supplied. As a result, we might consider retrieving _all_ the tasks now associated with the goal (such as with something like `all_task_ids = [task.id for task in goal.tasks]`), rather than returning a list only of the ids we just added.

If we interpretted this route as _replacing_ the tasks associated with this goal (such as by generating a list of the tasks, and then replacing the whole `goal.tasks` list with the updated list), then the current behavior of only returning the incoming tasks would more closely match expectations. This might match the behavior described above, where we might consider an empty list of task ids to mean there should be no tasks associated with the goal.

File: app/routes/goal_routes.py [hunk 28]
```diff
@@ -1 +1,145 @@
...
+def get_tasks_of_goal(goal_id):
+    goal = validate_goal(goal_id)
+    
+    tasks = [{
+        "id": task.id,
+        "goal_id": goal.id,
+        "title": task.title,
+        "description": task.description,
+        "is_complete": task.completed_at is not None
+    } for task in goal.tasks]
```
Comment (anselrognlie) [inline app/routes/goal_routes.py:138]: Notice that without a `to_dict` method, the conciseness of the list comprehension is lost. Compare to the case where we do create a `to_dict` method.
```py
    tasks = [task.to_dict() for task in goal.tasks]
```

We would need to be sure that the `to_dict` method correctly added the `goal_id` information only when available so that it could be used in all the places where we convert tasks to dictionary responses.

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
